File: 2023/09/solution.pas
{ solution.pas -- mirage maintenance -- Troy Brumley BlameTroi@gmail.com }


program solution;

{ expects to be compiled with the following two mode flags and command
line/fpc.cfg options of -Ciort -Sgix -Sh- for full checks, allowing
exceptions, goto, inlining, and prefer classic strings. }
{$longstrings off}            { to be sure, use AnsiString or pchar explicitly }
{$booleval on}                { standard pascal expects this }

type
  integer = int64;            { by default ints are 64 bits this decade}

{ advent of code 2023 day 9 -- mirage maintenance

  extrapolation of sensor readings using a pascal's triangle looking
  calculation. for part one extrapolate forward, and for part two
  extrapolate backward. }

const
{$define xxxTESTING }
{$define TESTONE }
{$ifdef TESTING }
   MAXSENSORS = 3;
   MAXREADINGS = 6;
{$else}
   MAXSENSORS = 200;
   MAXREADINGS = 21;
{$endif}
{$I constchars.p }


{ ---------------------------------------------------------
  global variables are appropriate for this problem. note
  that while sensors number from 1, the logs number from 0.
  the actual last slot is reserved for the new predicted
  reading.
  --------------------------------------------------------- }

var
  dataset    : array[1 .. MAXSENSORS] of string;
  sensorlogs : array[1 .. MAXSENSORS, 0 .. MAXREADINGS] of integer;
  scratchpad : array[0 .. MAXREADINGS, 0 .. MAXREADINGS] of integer;
  depthtrack : array[0 .. MAXREADINGS] of integer;

{ ---------------------------------------------------------
  utilitiy
  --------------------------------------------------------- }

{$I str2int.p }


{ zero the scratch pad. }

procedure zeroscratchpad;

var
   i, j : integer;

begin
   for i := low(scratchpad) to high(scratchpad) do
      for j := low(scratchpad[i]) to high(scratchpad[i]) do
         scratchpad[i, j] := 0;
end;


{ initialize global variables. }

procedure init;

var
   i, j : integer;

begin
   for i := low(dataset) to high(dataset) do
      dataset[i] := '';
   for i := low(sensorlogs) to high(sensorlogs) do
      for j := low(sensorlogs[i]) to high(sensorlogs[i]) do
         sensorlogs[i, j] := 0;
   for i := low(depthtrack) to high(depthtrack) do
      depthtrack[i] := 0;
   zeroscratchpad;
end;


{ is the whole row zero? and don't be cute! }

function scratchrowzero(r : integer) : boolean;

var
   j : integer;

begin
   scratchrowzero := false;
   for j := low(scratchpad[r]) to high(scratchpad[r]) do
      if scratchpad[r, j] <> 0 then
         exit;
   scratchrowzero := true;
end;


{$I aocinput.p }


{ ---------------------------------------------------------
  loader and support.
  --------------------------------------------------------- }

{ read the sensor logs and parse them into the matrix. }

procedure loadone;

var
   i, j : integer;
   p    : integer;

begin
   init;
   aocrewind;
   i := 0;
   while not aoceof do begin
      i := i + 1;
      dataset[i] := aocread;
      p := 1;
      j := 0;   { remember, this starts at 0 not 1 }
      while p < length(dataset[i]) do begin
         while dataset[i][p] = SPACE do
            p := p + 1;
         sensorlogs[i, j] := str2int(dataset[i], p);
         j := j + 1;
      end;
   end;
end;


{ loads the data in backwards }
procedure loadtwo;

var
   i, j : integer;
   p    : integer;

begin
   init;
   aocrewind;
   i := 0;
   while not aoceof do begin
      i := i + 1;
      dataset[i] := aocread;
      p := 1;
      { switch directions on load for part two }
      j := MAXREADINGS - 1;
      while p < length(dataset[i]) do begin
         while dataset[i][p] = SPACE do
            p := p + 1;
         sensorlogs[i, j] := str2int(dataset[i], p);
         j := j - 1;
      end;
   end;
end;


{ determine the next figure in the series by using the differences of
  the prior readings. the extrapolation uses a pascal's triangle like
  approach to the extrapolation. }

function extrapolate(s : integer): integer; { sensor number to process }

var
   i, j : integer;

begin
   { initialize scratchpad for current sensor. }
   zeroscratchpad;
   for j := low(scratchpad[0]) to high(scratchpad[0]) do
      scratchpad[0][j] := sensorlogs[s, j];

   { calculate differences, always looks back to prior line }
   { each row (i) needs one fewer column than the one above.
     the last column in the sensor log is 0 and that's the
     value we need to replace. }
   i := low(scratchpad); { sensor log }
   while i < high(scratchpad) do begin
      if scratchrowzero(i) then begin
         depthtrack[i] := depthtrack[i] + 1;
         //writeln(output, 'differences completed at ', i);
         break;
      end;
      i := i + 1;
      for j := low(scratchpad[i]) to high(scratchpad[i]) - i - 1 do
         scratchpad[i, j] := scratchpad[i-1, j+1] - scratchpad[i-1, j];
   end;
   { now work back up the pad replacing the zero difference at the
     end with the correct value. current row's tail + prior row's
     tail. }
   repeat
      j := high(scratchpad[i]) - i;
      scratchpad[i-1, j+1] := scratchpad[i, j] + scratchpad[i-1, j];
      i := i - 1;
   until i = low(scratchpad);
   extrapolate := scratchpad[low(scratchpad), high(scratchpad[0])];
end;


{ to check our work, extrapolate the next reading for each
  sensor and report the sum of the extrapolations. }

function partone : integer;
var
   i : integer;

begin
   loadone;
   for i := 1 to high(sensorlogs) do
      sensorlogs[i, high(sensorlogs[i])] := extrapolate(i);
   partone := 0;
   for i := 1 to high(sensorlogs) do
      partone := partone + sensorlogs[i, high(sensorlogs[i])];
end;


{ part two is the same as part one, but the loader inverts
  the data so we are actually extrapolating backward. }

function parttwo : integer;

var
   i : integer;

begin
   loadtwo;
   for i := 1 to high(sensorlogs) do
      sensorlogs[i, high(sensorlogs[i])] := extrapolate(i);
   parttwo := 0;
   for i := 1 to high(sensorlogs) do
      parttwo := parttwo + sensorlogs[i, high(sensorlogs[i])];
end;


{ mainline }

var
   i : integer;

begin
   init;
   aocopen;
   i := partone;
   writeln('part one: ', i);
   i := parttwo;
   writeln('part two: ', i);
   aocclose;
end.

File: 2023/09/solution.f90
! solution.f90 -- 2022 day 9 -- mirage maintenance -- troy brumley blametroi@gmail.com

! advent of code 2023 day 9 -- mirage maintenance
!
! extrapolation of sensor readings using a triangular structure that reminds me
! of pascal's triangle. for part one, extrapolate forward. for part two,
! extrapolate backward. the easiest way to do this is to invert the rows of
! readings on the load.

program solution

   use iso_fortran_env, only: int64
   use aocinput, only: open_aoc_input, read_aoc_input, close_aoc_input, max_aoc_reclen, rewind_aoc_input
   use aochelpers, only: str_int, str_ints

   implicit none

   integer, parameter            :: AOCIN = 10
   integer(kind=int64)           :: part_one, part_two  ! results
   character(len=max_aoc_reclen) :: rec

   integer, parameter :: MAXSENSORS = 200 ! 3 in small test set
   integer, parameter :: MAXREADINGS = 21 ! 6 in small set

   ! note that while sensors number from 1, the logs number from 0. the actual
   ! last slot is reserved for the new predicted reading.

   integer(kind=int64) :: sensorlogs(1:MAXSENSORS, 0:MAXREADINGS)
   integer             :: numsensors
   integer             :: numreadings
   integer(kind=int64) :: scratchpad(0:MAXREADINGS, 0:MAXREADINGS)

   ! initialize
   rec = ""
   call open_aoc_input(AOCIN)
   part_one = 0; part_two = 0

   call init
   call loadone
   part_one = dopartone()

   call init
   call loadtwo
   part_two = doparttwo()

   ! report and close
   print *
   print *, part_one, "part one"
   print *, part_two, "part two"
   print *

   call close_aoc_input(AOCIN)

contains

   ! initialize global variables.

   subroutine init
      implicit none

      rec = ""
      sensorlogs = 0
      scratchpad = 0
      numsensors = 0
      numreadings = 0
   end subroutine init

   ! read sensor logs and parse into the matrix

   subroutine loadone
      implicit none
      integer              :: i, j
      integer, allocatable :: nums(:)

      call rewind_aoc_input(AOCIN)
      i = 0
      do while (read_aoc_input(AOCIN, rec))
         if (len_trim(rec) < 1) cycle
         if (rec(1:1) == "#") cycle
         i = i + 1
         nums = str_ints(rec, " ")
         numreadings = size(nums)
         do j = 1, size(nums)
            sensorlogs(i, j - 1) = nums(j)
         end do
      end do
      numsensors = i
   end subroutine loadone

   ! load one but backwards

   subroutine loadtwo
      implicit none
      integer              :: i, j
      integer, allocatable :: nums(:)

      call rewind_aoc_input(AOCIN)
      i = 0
      numreadings = -1
      do while (read_aoc_input(AOCIN, rec))
         if (len_trim(rec) < 1) cycle
         if (rec(1:1) == "#") cycle
         i = i + 1
         nums = str_ints(rec, " ")
         numreadings = size(nums)
         ! switch directions on load for part two
         do j = 1, numreadings
            sensorlogs(i, j - 1) = nums(numreadings + 1 - j)
         end do
      end do
      numsensors = i
   end subroutine loadtwo

   ! to check our work, extrapolate the next reading for each
   ! sensor and report the sum of the extrapolations. }

   function dopartone() result(res)
      implicit none
      integer(kind=int64) :: res
      integer             :: i

      res = 0
      do i = 1, numsensors
         res = res + extrapolate(i)
      end do
   end function dopartone

   ! part two, ends up the same as part one

   function doparttwo() result(res)
      implicit none
      integer(kind=int64) :: res
      integer             :: i

      res = 0
      do i = 1, numsensors
         res = res + extrapolate(i)
      end do
   end function doparttwo

   ! determine the next figure in the series by using the differences of the
   ! prior readings. the extrapolation uses a pascal's triangle like structure
   ! for the extrapolation.

   function extrapolate(s) result(res)
      implicit none
      integer, intent(in) :: s ! sensor number
      integer(kind=int64) :: res
      integer             :: i, j

      ! load up scratchpad with current sensor

      scratchpad = 0
      do j = 0, numreadings
         scratchpad(0, j) = sensorlogs(s, j)
      end do

      ! calculate differences, always looks back to prior line.
      ! each row (i) needs one fewer column than the one above.
      ! the last column in the sensor log is 0 and that's the
      ! value we need to replace.

      i = 0
      do while (i < numreadings) ! not <=, last slot reserved

         ! if the prior row is all zero, we're done
         if (isscratchrowzero(i)) exit

         ! calculate this row from prior row , this = pripr + prior+1
         i = i + 1
         do j = 0, numreadings - i - 1
            scratchpad(i, j) = scratchpad(i - 1, j + 1) - scratchpad(i - 1, j)
         end do

      end do

      ! now work back up the pad replacing the zero difference at the
      ! end with the correct value. current row's tail + prior row's
      ! tail.

      do while (i /= 0)
         j = numreadings - i
         scratchpad(i - 1, j + 1) = scratchpad(i, j) + scratchpad(i - 1, j)
         i = i - 1
      end do
      res = scratchpad(0, numreadings)
   end function extrapolate

   ! is this whole row of scratch zero?
   ! and is there a better fortran way to do the check?

   function isscratchrowzero(r) result(res)
      implicit none
      integer, intent(in) :: r
      logical             :: res
      integer             :: i

      res = .false.
      do i = 0, MAXREADINGS
         if (scratchpad(r, i) /= 0) return
      end do
      res = .true.
   end function isscratchrowzero

end program solution

